test/e2e/framework/framework.go

```diff
@@ -0,0 +1,249 @@
+    "k8s.io/client-go/kubernetes/scheme"
+    "k8s.io/client-go/rest"
+
+    corev1 "k8s.io/api/core/v1"
+    v1 "k8s.io/api/core/v1"
+    apierrors "k8s.io/apimachinery/pkg/api/errors"
+    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+    kubeclientset "k8s.io/client-go/kubernetes"
+
+    . "github.com/onsi/gomega"
+)
+
+const (
+    // How long to try single API calls (like 'get' or 'list'). Used to prevent
+    // transient failures from failing tests.
+    DefaultSingleCallTimeout = 30 * time.Second
+    // Polling interval while trying to create objects
+    PollInterval = 100 * time.Millisecond
+)
+
+// Framework supports common operations used by e2e tests; it will keep a client & a namespace for you.
+// Eventual goal is to merge this with integration test framework.
+type Framework struct {
+    BaseName string
+
+    // Set together with creating the ClientSet and the namespace.
+    // Guaranteed to be unique in the cluster even when running the same
+    // test multiple times in parallel.
+    UniqueName string
+
+    ClientSet   *kubeclientset.Clientset
+
+    SkipNamespaceCreation    bool            // Whether to skip creating a namespace
+    Namespace                *v1.Namespace   // Every test has at least one namespace unless creation is skipped
+    namespacesToDelete       []*v1.Namespace // Some tests have more than one.
+    NamespaceDeletionTimeout time.Duration
+
+    // To make sure that this framework cleans up after itself, no matter what,
+    // we install a Cleanup action before each test and clear it after.  If we
+    // should abort, the AfterSuite hook should run all Cleanup actions.
+    cleanupHandle            CleanupActionHandle
+
+    // configuration for framework's client
+    Options Options
+
+}
+
+// Options is a struct for managing test framework options.
+type Options struct {
+    ClientQPS    float32
+    ClientBurst  int
+    GroupVersion *schema.GroupVersion
+}
+
+// NewDefaultFramework makes a new framework and sets up a BeforeEach/AfterEach for
+// you (you can write additional before/after each functions).
+func NewDefaultFramework(baseName string) *Framework {
+    options := Options{
+        ClientQPS:   20,
+        ClientBurst: 50,
+    }
+    return NewFramework(baseName, options, nil)
+}
+
+// NewFramework creates a test framework.
+func NewFramework(baseName string, options Options, client *kubeclientset.Clientset) *Framework {
+    f := &Framework{
+        BaseName:                 baseName,
+        Options:                  options,
+        ClientSet:                client,
+    }
+
+    ginkgo.BeforeEach(f.BeforeEach)
+    ginkgo.AfterEach(f.AfterEach)
+
+    return f
+}
+
+
+func (f *Framework) BeforeEach() {
+    // The fact that we need this feels like a bug in ginkgo.
+    // https://github.com/onsi/ginkgo/issues/222
+    f.cleanupHandle = AddCleanupAction(f.AfterEach)
+
+    if f.ClientSet == nil {
+        ginkgo.By("Creating a kubernetes client")
```

Comment:
(No action required) This block seems a fine candidate for encapsulation in a function.

Reply:
done 👍